pkg/pillar/cmd/baseosmgr/worker.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2020 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package baseosmgr
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/lf-edge/eve/pkg/pillar/worker"
+	"github.com/lf-edge/eve/pkg/pillar/zboot"
+)
+
+// installWorkDescription install work we feed into the worker go routine
+type installWorkDescription struct {
+	contentID string
+	ref       string
+	target    string
+}
+
+// installWorkResult result of sending to a partition
+type installWorkResult struct {
+	worker.WorkResult // Error etc
+}
+
+var pendingInstallMap = make(map[string]bool)
+var installWorkResultMap = make(map[string]installWorkResult)
+
+func lookupPendingInstall(key string) bool {
+	res, ok := pendingInstallMap[key]
+	return ok && res
+}
+
+func addPendingInstall(key string) {
+	pendingInstallMap[key] = true
+}
+
+func deletePendingInstall(key string) {
+	delete(pendingInstallMap, key)
+}
+
+// AddWorkInstall checks if the Key is in the map of pending work
+// and if not kicks of a worker and adds it
+func AddWorkInstall(ctx *baseOsMgrContext, key, ref, target string) {
+	log.Infof("AddWorkInstall(%s)", key)
+	if lookupPendingInstall(key) {
+		log.Infof("AddWorkInstall(%s) found", key)
+		return
+	}
+	d := installWorkDescription{
+		contentID: key,
+		ref:       ref,
+		target:    target,
+	}
+	w := worker.Work{Key: key, Description: d}
+	ctx.worker.Submit(w)
+	addPendingInstall(key)
+	log.Infof("AddWorkInstall(%s) done", key)
+}
+
+// DeleteWorkInstall is called by user when work is done
+func DeleteWorkInstall(key string) {
+	log.Infof("DeleteWorkInstall(%s)", key)
+	if !lookupPendingInstall(key) {
+		log.Infof("DeleteWorkInstall(%s) NOT found", key)
+		return
+	}
+	deletePendingInstall(key)
+	log.Infof("DeleteWorkInstall(%s) done", key)
+}
+
+func lookupInstallWorkResult(key string) *installWorkResult {
+	if res, ok := installWorkResultMap[key]; ok {
+		return &res
+	}
+	return nil
+}
+
+func addInstallWorkResult(key string, res installWorkResult) {
+	installWorkResultMap[key] = res
+}
+
+func deleteInstallWorkResult(key string) {
+	delete(installWorkResultMap, key)
+}
+
+// HandleWorkResult processes what comes out of the select loop
+func HandleWorkResult(ctx *baseOsMgrContext, res worker.WorkResult) {
+	// we do not really need a switch here, but we might have more types in the future
+	switch res.Description.(type) {
+	case installWorkDescription:
+		processInstallWorkResult(ctx, res)
+	}
+}
+
+// WorkerHandler worker switchboard for different types of workers
+func WorkerHandler(ctxPtr interface{}, w worker.Work) worker.WorkResult {
+	// we do not really need a switch here, but we might have more types in the future
+	switch t := w.Description.(type) {
+	case installWorkDescription:
+		return installWorker(ctxPtr, w)
+	default:
```

Comment:
Why is this in a default case? Is it for the case when the Description has not yet been initialized??

Reply:
No, it is for when it receives an unknown type. This always has to return a `worker.WorkResult`, so it returns one with an error. Is there a better option?

Reply:
I would fatal in that case, since that shouldn't happen.